File: agenkit-cs/src/Agenkit/Composition/ParallelAgent.cs
using System.Text;
using Agenkit.Core;

namespace Agenkit.Composition;

/// <summary>
/// Runs multiple agents concurrently and aggregates their responses.
/// </summary>
public class ParallelAgent : IAgent
{
    private readonly IReadOnlyList<IAgent> _agents;
    private readonly Func<IReadOnlyList<Message>, Message>? _aggregator;

    /// <inheritdoc />
    public string Name => "ParallelAgent";

    /// <inheritdoc />
    public IReadOnlyList<string> Capabilities => new[] { "parallel", "fan-out" };

    /// <summary>
    /// Creates a new ParallelAgent.
    /// </summary>
    /// <param name="agents">Agents to run concurrently.</param>
    /// <param name="aggregator">Optional custom aggregation function. Defaults to concatenation.</param>
    public ParallelAgent(IReadOnlyList<IAgent> agents, Func<IReadOnlyList<Message>, Message>? aggregator = null)
    {
        if (agents.Count == 0)
            throw new ArgumentException("at least one agent is required", nameof(agents));
        _agents = agents;
        _aggregator = aggregator;
    }

    /// <inheritdoc />
    public async Task<Message> ProcessAsync(Message message, CancellationToken ct = default)
    {
        var tasks = _agents.Select(a => a.ProcessAsync(message, ct)).ToList();
        var results = await Task.WhenAll(tasks).ConfigureAwait(false);

        return _aggregator is not null
            ? _aggregator(results)
            : DefaultAggregate(results);
    }

    /// <inheritdoc />
    public IntrospectionResult Introspect() => new(
        Name,
        Capabilities,
        State: new Dictionary<string, object> { ["agent_count"] = _agents.Count });

    private static Message DefaultAggregate(IReadOnlyList<Message> results)
    {
        var sb = new StringBuilder();
        for (int i = 0; i < results.Count; i++)
        {
            if (i > 0) sb.AppendLine();
            sb.Append(results[i].ContentString());
        }
        return Message.NewMessage("assistant", sb.ToString());
    }
}

File: agenkit-scala/src/main/scala/io/agenkit/composition/ParallelAgent.scala
package io.agenkit.composition

import io.agenkit.core.*
import scala.concurrent.{ExecutionContext, Future}
import java.util.concurrent.atomic.AtomicLong

/** Runs all agents concurrently and merges their responses with a configurable strategy. */
class ParallelAgent(
  val name: String,
  agents: List[Agent],
  merger: List[Message] => Message = msgs =>
    Message.of("assistant", msgs.map(_.contentString).mkString("\n\n"))
) extends Agent:
  private val _processedCount = new AtomicLong(0)

  def capabilities: List[String] = List("parallel", "fan-out")

  def process(message: Message)(using ExecutionContext): Future[Message] =
    _processedCount.incrementAndGet()
    Future.sequence(agents.map(_.process(message))).map(merger)

  def introspect(): IntrospectionResult =
    IntrospectionResult(
      name = name,
      metadata = Map("agentCount" -> agents.size, "agents" -> agents.map(_.name)),
      capabilities = capabilities,
      processedMessages = _processedCount.get()
    )
